### DISCUSSION TITLE
Fixing problems with fastq-dump

### DISCUSSION THREAD

**Discussion by ialbert** [biostars/biostar-handbook-issues#239] [Announcements]
Some reader report various errors with fastq-dump that look like:

```
Failed to create TLS stream for 'www.ncbi.nlm.nih.gov' (130.14.29.110)
item not found while constructing within virtual database module - the path 'SRR21675623' cannot be opened as database or table 
```

or 

```
( !! The certificate is not correctly signed by the trusted CA !! The certificate is signed with an unacceptable hash. )
```

In those cases, we found that installing a newer version of `sra-tools` fixes the problem.

We have an installer that can download and install the native client distributed by NCBI.  Run it like so:

    curl http://data.biostarhandbook.com/sratools.sh | bash

After running the above, close and reopen the terminal, then test the results with:

    fastq-dump --stdout -X 2 SRR390728

I believe the above is a solution to the problems people have been experiencing.

**Comment by ChronoWeaver** [comment]
I tried again according to your method, and running "fastq-dump --stdout -X 2 SRR390728" succeeded, but when I run "fastq-dump -X 1 -Z SRR1553591" directly, it will report an error, I don't know the reason for the error.(M1 macbook pro)

$ fastq-dump --stdout -X 2 SRR390728

Read 2 spots for SRR390728
Written 2 spots for SRR390728
@SRR390728.1 1 length=72
CATTCTTCACGTAGTTCTCGAGCCTTGGTTTTCAGCGATGGAGAATGACTTTGACAAGCTGAGAGAAGNTNC
+SRR390728.1 1 length=72
;;;;;;;;;;;;;;;;;;;;;;;;;;;9;;665142;;;;;;;;;;;;;;;;;;;;;;;;;;;;;96&&&&(
@SRR390728.2 2 length=72
AAGTAGGTCTCGTCTGTGTTTTCTACGAGCTTGTGTTCCAGCTGACCCACTCCCTGGGTGGGGGGACTGGGT
+SRR390728.2 2 length=72
;;;;;;;;;;;;;;;;;4;;;;3;393.1+4&&5&&;;;;;;;;;;;;;;;;;;;;;<9;<;;;;;464262

$ fastq-dump -X 1 -Z SRR1553591
2022-10-23T02:33:37 fastq-dump.3.0.0 sys: encryption failed while validating token within cryptographic module - Verification issue 0x4008 for this certificate: (  cert. version     : 3  serial number     : 00  issuer name       : C=US, O=The Go Daddy Group, Inc., OU=Go Daddy Class 2 Certification Authority  subject name      : C=US, O=The Go Daddy Group, Inc., OU=Go Daddy Class 2 Certification Authority  issued  on        : 2004-06-29 17:06:20  expires on        : 2034-06-29 17:06:20  signed using      : RSA with SHA1  RSA key size      : 2048 bits  basic constraints : CA=true  )
...
2022-10-23T02:33:37 fastq-dump.3.0.0 sys: mbedtls_ssl_get_verify_result for 'locate.ncbi.nlm.nih.gov' returned 0x4008 ( !! The certificate is not correctly signed by the trusted CA  !! The certificate is signed with an unacceptable hash. )
Failed to call external services.
(bioinfo)

**Comment by ialbert** [comment]
oh, just realized that this was the same thread,

this sounds like some sort of network problem frankly - lots of certificate problems

**Comment by ialbert** [comment]
A workaround would be to 

    bio search SRR1553591

that prints
```

[
    {
        "run_accession": "SRR1553591",
        "sample_accession": "SAMN02952040",
        "first_public": "2015-06-05",
        "country": "Sierra Leone",
        "sample_alias": "G3846",
        "fastq_bytes": "50369239;52834496",
        "read_count": "789171",
        "library_name": "G3846_r1.ADXX",
        "library_strategy": "RNA-Seq",
        "library_source": "TRANSCRIPTOMIC",
        "library_layout": "PAIRED",
        "instrument_platform": "ILLUMINA",
        "instrument_model": "Illumina HiSeq 2500",
        "study_title": "Zaire ebolavirus Genome sequencing",
        "fastq_ftp": "ftp.sra.ebi.ac.uk/vol1/fastq/SRR155/001/SRR1553591/SRR1553591_1.fastq.gz;ftp.sra.ebi.ac.uk/vol1/fastq/SRR155/001/SRR1553591/SRR1553591_2.fastq.gz"
    
```    

you can download from those links (automate the whole thing with `jq`)

**Reply by ChronoWeaver** [reply]
thank you for your reply。
I used your method ，using "$ wget ftp.sra.ebi.ac.uk/vol1/fastq/SRR155/001/SRR1553591/SRR1553591_1.fastq.gz;ftp.sra.ebi.ac.uk/vol1/fastq /SRR155/001/SRR1553591/SRR1553591_2.fastq.gz" downloaded the corresponding file, what do I need to do next

**Reply by ChronoWeaver** [reply]
Can you give me a more detailed explanation, I just started using your book to learn, I don't know how to use the code very well, I am very grateful.

**Reply by ChronoWeaver** [reply]
[https://github.com/biostars/biostar-handbook/discussions/200#discussioncomment-3310589](https://github.com/biostars/biostar-handbook/discussions/200#discussioncomment-3310589)
Thanks again, I succeeded using the manual setup method below in this question.

**Comment by cborowski** [comment]
I need help with fastq-dump errors after Biostar Handbook software installation. I am not very experienced and frankly I don’t understand the errors. Installation went well so far, but doctor.py returns: 
fastq-dump -X 1 -Z SRR1553591 ... ERROR

After running 
curl http://data.biostarhandbook.com/sratools.sh | bash
+ terminal reopening and 
fastq-dump --stdout -X 2 SRR390728
I get: 

2022-12-22T10:42:21 fastq-dump.2.9.6 sys: connection failed while opening file within cryptographic module - mbedtls_ssl_handshake returned -9984 ( X509 - Certificate verification failed, e.g. CRL, CA or signature check failed )
2022-12-22T10:42:21 fastq-dump.2.9.6 sys: mbedtls_ssl_get_verify_result returned 0x4008 (  !! The certificate is not correctly signed by the trusted CA  !! The certificate is signed with an unacceptable hash.  )
2022-12-22T10:42:21 fastq-dump.2.9.6 sys: connection failed while opening file within cryptographic module - ktls_handshake failed while accessing '130.14.29.110' from '192.168.2.120'
2022-12-22T10:42:21 fastq-dump.2.9.6 sys: connection failed while opening file within cryptographic module - Failed to create TLS stream for 'www.ncbi.nlm.nih.gov' (130.14.29.110) from '192.168.2.120'
2022-12-22T10:42:21 fastq-dump.2.9.6 err: item not found while constructing within virtual database module - the path 'SRR390728' cannot be opened as database or table
(bioinfo)
 
Running
bio search SRR1553591
returns 
[
    {
        "run_accession": "SRR1553591",
        "sample_accession": "SAMN02952040",
        "first_public": "2015-06-05",
        "country": "Sierra Leone",
        "sample_alias": "G3846",
        "fastq_bytes": "50369239;52834496",
        "read_count": "789171",
        "library_name": "G3846_r1.ADXX",
        "library_strategy": "RNA-Seq",
        "library_source": "TRANSCRIPTOMIC",
        "library_layout": "PAIRED",
        "instrument_platform": "ILLUMINA",
        "instrument_model": "Illumina HiSeq 2500",
        "study_title": "Zaire ebolavirus Genome sequencing",
        "fastq_ftp": "ftp.sra.ebi.ac.uk/vol1/fastq/SRR155/001/SRR1553591/SRR1553591_1.fastq.gz;ftp.sra.ebi.ac.uk/vol1/fastq/SRR155/001/SRR1553591/SRR1553591_2.fastq.gz"
    }
]
(bioinfo)

What does that mean? I understand that the ftp site is the download location for the ebola data, but the fastq-dump tool is still not working. Should i ignore it?

**Reply by ialbert** [reply]
It is not your fault for being a beginner.

What happens is that the people that run NCBI cannot get their stuff together to the point of absurditity that downloading a file is too complicated ... 

Just ponder and think about that for a second ... to truly understand the magnitude of them letting an entire domain of science down

The Biostar Workflows explains the process in more detail on what to do when fastq-dump fails, see the section: When fastq-dump fails.

https://www.biostarhandbook.com/books/workflows/tutorials/fastq-data/#when-fastq-dump-fails

**Reply by ialbert** [reply]
In a nutshell you can download the links yourself:

    wget ftp.sra.ebi.ac.uk/vol1/fastq/SRR155/005/SRR1553425/SRR1553425_1.fastq
    wget ftp.sra.ebi.ac.uk/vol1/fastq/SRR155/005/SRR1553425/SRR1553425_2.fastq

**Reply by yonglu1** [reply]
I had the same problem after run: curl http://data.biostarhandbook.com/sratools.sh | bash
I believe there is something wrong in the Bioinfo environment. I can run fastq-dump without a problem before activate Bioinfo. Please refer to the attached picture for detailed information.
![Screenshot from 2023-01-18 19-31-18](https://user-images.githubusercontent.com/14941689/213330721-66db3ed0-5cfc-4a62-9683-c766e37ee744.png)

**Reply by ialbert** [reply]
that means you have two different versions of fastq-dump installed and the system finds the broken one first

uninstall sratools from the bioinfo environment or move the other fastq-dump to be earlier in the PATH variable


    mamba uninstall sra-tools

**Reply by yonglu1** [reply]
Thank you so much for the prompt reply. uninstall sratools from the bioinfo environment fixed my problem. Thanks!

**Comment by cborowski** [comment]
Man thanks for your quick response, Istvan.

I had gone through the When-fastq-dump-fails section several times, but 
I did not manage to extrapolate the right solution for. At the end I 
tried and tried and suddenly several steps of the installation and file 
retrieving process worked that had not done so before. No clue why, but 
at the moment I am fine.

Nevertheless, 1000 thanks for your immediate help - this is really great 
support!
Christian



I On 22/12/2022 23:41, Istvan Albert wrote:

> It is not your fault for being a beginner.
>
> What happens is that the people that run NCBI cannot get their stuff 
> together to the point of absurditity that downloading a file is too 
> complicated ...
>
> Just ponder and think about that for a second ... to truly understand 
> the magnitude of letting a entire domain of science down
>
> The Biostar Workflows explains the process in more detail on what to 
> do when fastq-dump fails, see the section: When fastq-dump fails.
>
> https://www.biostarhandbook.com/books/workflows/tutorials/fastq-data/#when-fastq-dump-fails
>
> —
> Reply to this email directly, view it on GitHub 
> <https://github.com/biostars/biostar-handbook/discussions/239#discussioncomment-4479944>, 
> or unsubscribe 
> <https://github.com/notifications/unsubscribe-auth/AHNHEAND2V7UHASSKD4TFRDWOTKKDANCNFSM6AAAAAAQX6BVGY>.
> You are receiving this because you commented.Message ID: 
> ***@***.***>
>
-- 
	Dr. Christian Borowski

Department of Symbiosis

Max Planck Institute for Marine Microbiology
Celsiusstr. 1 - D-28359 Bremen - Room 3242
Phone: +49 421 2028 823
eMail: ***@***.***

**Comment by nicolemadrid** [comment]
<img width="587" alt="Screenshot 2023-01-22 at 7 35 08 PM" src="https://user-images.githubusercontent.com/123338721/213949119-f8549135-bbcc-453f-bfa3-66f694f21d10.png">
I followed the directions in the 'How to fix fastq-dump errors' section but keep receiving this message, I am not familiar with coding and am following this textbook for a college course. I would be grateful for any help.

**Reply by ialbert** [reply]
uninstall the default sra-tools package

    mamba uninstall sra-tools

then close and reopen the terminal

**Reply by nicolemadrid** [reply]
<img width="939" alt="Screenshot 2023-01-22 at 8 55 36 PM" src="https://user-images.githubusercontent.com/123338721/213954065-fb239cc2-a73a-46bc-9271-9e2b1d27e4f5.png">

I uninstalled the default sra-tools package, closed and reopened the terminal and this is the result following the how to fix fastq-dump errors.

**Reply by ialbert** [reply]
there are a number of similar issues that people experience:


https://github.com/ncbi/sra-tools/issues/598

and

https://github.com/ncbi/sra-tools/issues/467

and a number of others. Perhaps it is a network problem, it sounds like it.

If you keep having problems look at the section "When fastq-dump fails" here:

https://www.biostarhandbook.com/books/workflows/tutorials/fastq-data/#when-fastq-dump-fails

to locate the URLs then download those URLs with `wget` or `curl`

**Comment by davincios** [comment]
I love the doctor.py script -- super helpful. 

But I can confirm that I doctory.py has fastq-dump errors for me too. 

"curl http://data.biostarhandbook.com/sratools.sh | bash" 

does not solve the issue. 

I am using a macbook air M1.

**Reply by davincios** [reply]
Following chat gpt does work:


![image](https://github.com/biostars/biostar-handbook/assets/33206282/0c527548-8821-4ff2-a309-752ce2bbc229)

**Reply by davincios** [reply]
fastq-dump works when exporting to the correct path.

**Reply by ialbert** [reply]
thanks for posting the answer

the install protocols changed slightly, 

the fix you note is should be necessary only  if someone used the install instructions from an earlier version of the book (March 2023).

Since then the book specific bash config has been placed in a separate file to keep these settings separate from other settings.

long story short current installations should not need any other fix, but the fix you posted is necessary if someone already has an working installation and then sratools goes bad and they try the solution I posted above.

**Reply by davincios** [reply]
I think I declared victory prematurely because everytime I reopen the terminal I have to go through the same chores again. I have now removed conda and reinstalled it, as well as running the curl script to install the bioinformatics software again.

**Reply by davincios** [reply]
For future reference here are the steps I have taken before:

**I am getting the bioinfo error after conda activate bioinfo**
conda activate bioinfo // Could not find conda environment: bioinfo
—----------------------
conda info --envs
# conda environments:
                         /Users/vincent/miniconda3
                         /Users/vincent/miniconda3/envs/bioinfo
base                  *  /Users/vincent/opt/anaconda3

How to solve:
conda activate /Users/vincent/miniconda3/envs/bioinfo

Then run doctor.py by going to bin
cd bin 
python doctor.py 
// results in:
# fastq-dump    --> ERROR
# fastq-dump -X 1 -Z SRR1553591 --> ERROR

**Adding fastq-dump separately**
curl http://data.biostarhandbook.com/sratools.sh | bash
cd sratoolkit.3.0.0-mac64/bin
pwd 
export PATH="/Users/vincent/bin/sratoolkit.3.0.0-mac64/bin:$PATH"
—--
fastq-dump --version
// returns "fastq-dump" version 3.0.0

Add to path with zsh
source ~/.zshrc

**Then run doctor.py again**
cd .. 
cd .. 
python doctor.py

returns:


**Add bin to path**
export PATH="/Users/vincent/bin:$PATH"

**Attempt to install software again**
curl http://data.biostarhandbook.com/install.sh | bash

**Error source ~/.bash_profile**
source ~/.bash_profile
// returns: /Users/vincent/.biostar.sh:12: = not found
source ~/.zsh_profile 

**Try adding conda again**
conda activate /Users/vincent/miniconda3/envs/bioinfo

**Add fastq-dump path again**
export PATH="/Users/vincent/bin/sratoolkit.3.0.0-mac64/bin:$PATH"

**Do doctor.py again**
![image](https://github.com/biostars/biostar-handbook/assets/33206282/b262f88a-4190-4bb6-86e3-975caa105429)

**Attempt to install software again**
curl http://data.biostarhandbook.com/install.sh | bash

**Error source ~/.bash_profile**
source ~/.bash_profile
// returns: /Users/vincent/.biostar.sh:12: = not found
source ~/.zsh_profile 

**Try adding conda again**
conda activate /Users/vincent/miniconda3/envs/bioinfo

**Add fastq-dump path again**
export PATH="/Users/vincent/bin/sratoolkit.3.0.0-mac64/bin:$PATH"

**Do doctor.py again**
![image](https://github.com/biostars/biostar-handbook/assets/33206282/3b6d590b-f70b-41d5-90e0-d3044deac43f)

**The next time I open terminal it doesn’t work aymore**
The next time I open terminal I have to go through the whole sequence again to make it work.

**Reply by davincios** [reply]
This is what happends the next time I open terminal:

<img width="568" alt="image" src="https://github.com/biostars/biostar-handbook/assets/33206282/6e5fe1e8-7362-4d99-b5ba-94edee9199d6">

**Reply by davincios** [reply]
@ialbert thank you for getting back to me previously!

I am just at a loss right now as I have been struggling now for hours.

**Reply by ialbert** [reply]
it appears that you have both anaconda and miniconda installed. 

basically that means you probably have two `condas` and you have to use the "right" conda 

instead of just typing conda try the full path, probably something like

~/miniconda3/bin/conda

might not be sufficient,

but if you already have anaconda, you would not need miniconda as well (the two things are the same technically, one much smaller than the other)

**Reply by davincios** [reply]
What are you suggesting? Should I remove miniconda?

**Reply by davincios** [reply]
if I understand correctly:

curl http://data.biostarhandbook.com/install.sh | bash

installs miniconda

**Reply by ialbert** [reply]
Perhaps go to the step by step instructions and continue with the step 5, after conda is installed. basically you don't need miniconda at all since you have anaconda (which is the same thing)

https://www.biostarhandbook.com/step-by-step.html

I am hoping that doing this will make your anaconda work like miniconda.

**Reply by davincios** [reply]
Ok thank you. I will try that.

**Reply by davincios** [reply]
No luck. Same issues.

**Reply by davincios** [reply]
Going to try on an AWS EC2 instance. Hopefully that will work

**Reply by davincios** [reply]
Same issues again with fasterq-dump on AWS EC2 as on MacOS:

<img width="1149" alt="image" src="https://github.com/biostars/biostar-handbook/assets/33206282/fa65053d-3faf-49f3-95ae-d8d9fb4ef931">

@ialbert

**Reply by ialbert** [reply]
PATH problems are the simplest but initially also the most frustrating issues. There is a dedicated chapter to PATH alone:

https://www.biostarhandbook.com/bashpath.html

in this case I believe the sratools installer should be under:

    ~/bin/sratools-3.0.0/bin

look inside that directory, and you if the tool is there then really it is only the PATH that is not set properly.

try running that, then print your PATH and see what is up there.

Key point to remember:

> The PATH to any righteous tool is beset on all sides by the inequities of UNIX and the tyranny of Environment Variables. Blessed is he who, in the name of charity and good will, shepherds the weak through the valley of darkness, for he is truly his brother’s keeper and the finder of lost programs.

I have not run the installer in a while, but I will test it tomorrow, to make sure it still functions as it should.

**Reply by davincios** [reply]
> ```
> ~/bin/sratools-3.0.0/bin
> ```

Thank you very much. 

But the tool is not there:

<img width="1149" alt="image" src="https://github.com/biostars/biostar-handbook/assets/33206282/86de5592-dd40-4809-99e0-e219147b1e4f">

**Reply by davincios** [reply]
Finally solved it on AWS EC2 instance m5.large (8 GiB RAM + 2 vCPU) (UBUNTU) 120GB EBS

$sudo apt install sra-toolkit

<img width="1149" alt="image" src="https://github.com/biostars/biostar-handbook/assets/33206282/94215940-bbb5-4535-b69f-51cab87c9f72">

**Reply by ialbert** [reply]
keep looking don't give up so easy

    ls ~/bin

I forgot the exact pathname

**Reply by davincios** [reply]
Yes it is working now :) thank you 

But it does require installing fastq-dump separately

**Reply by davincios** [reply]
I do want to say that I love the Biostar handbooks so far. It probably has an incredible (indirect) influence on the world by equipping people with bioinformatics know how and enabling them to achieve good things for humanity.

**Reply by ialbert** [reply]
thank you for the kind words. The handbook has been conceived with many goals and tries to educate on multiple levels, and I am happy to see that come through.

**Reply by davincios** [reply]
@ialbert 

I can confirm that the installation worked cleanly today in one try on a new Ubuntu instance on AWS. 

It didn't work cleanly on AWS Linux / CentOS Linux or Mac OS M1. Perhaps indeed due to conflicting / earlier installed software. 

<img width="809" alt="image" src="https://github.com/biostars/biostar-handbook/assets/33206282/46f6962e-ac69-479a-bb3a-730581e6df85">

**Reply by ialbert** [reply]
good to know @davincios Thanks for following up.

**Comment by ayoelegs** [comment]
Hello,

I am receiving some errors. I have followed every prescription in the book. Kindly assist.

```
ayoelegs@James ~
$ doctor.py
# Doctor! Doctor! Give me the news.
# Checking symptoms ...
# bwa           --> ERROR
# datamash      --> ERROR
# fastqc -h     --> ERROR
# hisat2        --> ERROR
# seqret -h     --> ERROR
# featureCounts --> ERROR
# efetch        ... OK
# esearch       ... OK
# samtools      --> ERROR
# fastq-dump    ... OK
# bowtie2       --> ERROR
# bcftools      ... OK
# seqtk         --> ERROR
# seqkit        --> ERROR
# bio           --> ERROR
# fastq-dump -X 1 -Z SRR1553591 ... OK
#
# Found some problems
#
# Unable to run: bwa, datamash, fastqc -h, hisat2, seqret -h, featureCounts, samtools, bowtie2, seqtk, seqkit, bio
#
# Check the Biostar Handbook for installation help.

```

**Reply by ialbert** [reply]
I believe that the problem is  that you have not activated the `bioinfo` environment.

Make sure to do so. Close and open a new terminal, then activate the environment with

    micromamba activate bioinfo
